### ISSUE TITLE
Having require minitest/autorun in the test_helper causes duplicate runs

### ISSUE THREAD

**Issue by grosser** [rails/rails#25048]
Seems like rails is doing it's own runner now ... maybe set Minitest @@at_exit_hook_installed so it does not set another one ? (solved for me / tracked it down, but could be an annoying thing for other upgraders ...)

### Steps to reproduce

add require 'minitest/autorun' to test_helper

### Expected behavior
1 result printed or nice error message

### Actual behavior
duplicated output because tests are run twice

### System configuration
**Rails version**: 5.0.0.rc1

**Comment by kaspth** [comment]
Rails handles that for you with `rails/test_help` (which has included its own autorun setup since Rails 4: https://github.com/rails/rails/commit/5da4d5142a60edda0cb78071553993ee33bb8637).

Can you explain why you're requiring it? 😁

Closing for now.

**Comment by grosser** [comment]
only required it by accident / because it was in there for a long time and did not cause any issues ... the problem is solved for me/my app ... just wanted to let you know in case there is a simple workaround or some nice way to show that this will not work ... tracking it down to the duplicate autorun took me a few minutes

**Comment by kaspth** [comment]
Interesting! Are you running your tests with `bin/rake test` or `bin/rails test`? Do the tests run twice with both of those commands or just one?

I have a feeling `bin/rake` will make them run twice if you require autorun: https://github.com/rails/rails/blob/b98c1f948903cef4bdd88a8d6777e480f5770d2f/railties/lib/rails/test_unit/testing.rake#L15

**Comment by prathamesh-sonpatki** [comment]
I had same issue of tests running two times in CodeTriage with `bin/rails test`. Removing `require 'minitest/autorun` solved the issue.

**Comment by kaspth** [comment]
Far as I can tell requiring `minitest/autorun` in `test_helper` is not something we explicitly supported before, it just happened to work. @prathamesh-sonpatki are you interested in investigating this some more? To figure out why tests run 2 times all of a sudden?

**Comment by prathamesh-sonpatki** [comment]
Sure. I am travelling today. Can look into it tomorrow.

On Wednesday 18 May 2016, Kasper Timm Hansen <notifications@github.com>
wrote:

> Far as I can tell requiring minitest/autorun in test_helper is not
> something we explicitly supported before, it just happened to work.
> @prathamesh-sonpatki <https://github.com/prathamesh-sonpatki> are you
> interested in investigating this some more? To figure out why tests run 2
> times all of a sudden?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/issues/25048#issuecomment-220088925>
>


-- 
./prathamesh

 GITHUB <https://github.com/prathamesh-sonpatki> | | BigBinary
<https://BigBinary.com>

**Comment by mrteera** [comment]
I need to use `minitest-focus` then I try to put `require 'minitest/autorun'` into `test_helper.rb`.
It works but `rails test` execute twice instead once. (I have 8 tests)

```
# Running tests with run options --seed 25532:
........

Finished tests in 1.936486s, 4.1312 tests/s, 3.6148 assertions/s.
8 tests, 7 assertions, 0 failures, 0 errors, 0 skips

# Running tests with run options --seed 25532:
.

Finished tests in 0.007167s, 1255.7398 tests/s, 1116.2131 assertions/s.
9 tests, 8 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by arthurnn** [comment]
The issue here is that our [command](https://github.com/rails/rails/blob/88ff3f820aa88b7235b7554668f91a73532726af/railties/lib/rails/commands/test.rb) calls `Minitest.run(ARGV)` and minitest autorun [also calls that](https://github.com/seattlerb/minitest/blob/f95ef007ec64d956d88fedc37fd2ed7e106b777e/lib/minitest.rb#L62). So that means we have 2 `at_exit` hooks registered, and both are calling `Minitest.run`

Not sure what would be the solution here. maybe we could require 'minitest/autorun' instead so it would not get required twice.

**Comment by zenspider** [comment]
@arthurnn I don't think you have 2 `at_exit` handlers registered. Looks like your command isn't registering one and is calling `exit` + `Minitest.run` directly. You could check and/or set `@@installed_at_exit` in `Minitest` to avoid collision with `Minitest.autorun`.

**Comment by arthurnn** [comment]
should be fixed by #26515
